src/Microsoft.Health.Fhir.Core/Features/Operations/Import/ImportResourceLoader.cs

```diff
@@ -0,0 +1,180 @@
+
+                    while (processingTasks.Count >= MaxConcurrentCount)
+                    {
+                        if (cancellationToken.IsCancellationRequested)
+                        {
+                            throw new OperationCanceledException();
+                        }
+
+                        IEnumerable<ImportResource> importResources = await processingTasks.Dequeue();
+                        foreach (ImportResource importResource in importResources)
+                        {
+                            await outputChannel.Writer.WriteAsync(importResource, cancellationToken);
+                        }
+                    }
+
+                    processingTasks.Enqueue(ParseImportRawContentAsync(resourceType, buffer.ToArray(), sequenceIdGenerator));
+                    buffer.Clear();
+                }
+
+                processingTasks.Enqueue(ParseImportRawContentAsync(resourceType, buffer.ToArray(), sequenceIdGenerator));
+                while (processingTasks.Count > 0)
+                {
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw new OperationCanceledException();
+                    }
+
+                    IEnumerable<ImportResource> importResources = await processingTasks.Dequeue();
+                    foreach (ImportResource importResource in importResources)
+                    {
+                        await outputChannel.Writer.WriteAsync(importResource, cancellationToken);
+                    }
+                }
+
+                _logger.LogInformation($"{currentIndex} lines loaded.");
+            }
+            finally
+            {
+                outputChannel.Writer.Complete();
+
+                if (!string.IsNullOrEmpty(leaseId))
+                {
+                    await _integrationDataStoreClient.TryReleaseLeaseAsync(new Uri(resourceLocation), leaseId, cancellationToken);
+                }
+
+                _logger.LogInformation("Load resource from store complete.");
+            }
+        }
+
+        private async Task<IEnumerable<ImportResource>> ParseImportRawContentAsync(string resourceType, (string content, long index)[] rawContents, Func<long, long> idGenerator)
+        {
+            return await Task.Run(() =>
+            {
+                List<ImportResource> result = new List<ImportResource>();
+
+                foreach ((string content, long index) in rawContents)
+                {
+                    long id = idGenerator(index);
+
+                    try
+                    {
+                        ImportResource importResource = _importResourceParser.Parse(id, index, content);
+
+                        if (!string.IsNullOrEmpty(resourceType) && !resourceType.Equals(importResource.Resource?.ResourceTypeName, StringComparison.Ordinal))
+                        {
+                            throw new FormatException("Resource type not match.");
+                        }
```

Comment:
For my context, why do we need to do this? Are there optimizations applied based on resourceType?

Reply:
We want to do batch import for same partition which now is for resource type. Customer provide same type resources can achieve better performance.